x-pack/plugins/enterprise_search/public/applications/app_search/components/result_settings/logic/helpers.ts

```diff
@@ -0,0 +1,127 @@
+} from '../types';
+
+const updateAllFields = (
+  fields: FieldResultSettingObject | ServerFieldResultSettingObject,
+  newValue: FieldResultSetting | {}
+) => {
+  return Object.keys(fields).reduce(
+    (acc, fieldName) => ({ ...acc, [fieldName]: { ...newValue } }),
+    {}
+  );
+};
+
+export const clearAllServerFields = (fields: ServerFieldResultSettingObject) =>
+  updateAllFields(fields, {});
+
+export const clearAllFields = (fields: FieldResultSettingObject) => updateAllFields(fields, {});
+
+export const resetAllServerFields = (fields: ServerFieldResultSettingObject) =>
+  updateAllFields(fields, { raw: {} });
+
+export const resetAllFields = (fields: FieldResultSettingObject) =>
+  updateAllFields(fields, DEFAULT_FIELD_SETTINGS);
+
+export const convertToServerFieldResultSetting = (fieldResultSetting: FieldResultSetting) => {
+  const serverFieldResultSetting: ServerFieldResultSetting = {};
+  if (fieldResultSetting.raw) {
+    serverFieldResultSetting.raw = {};
+    if (fieldResultSetting.rawSize) {
+      serverFieldResultSetting.raw.size = fieldResultSetting.rawSize;
+    }
+  }
+
+  if (fieldResultSetting.snippet) {
+    serverFieldResultSetting.snippet = {};
+    if (fieldResultSetting.snippetFallback) {
+      serverFieldResultSetting.snippet.fallback = fieldResultSetting.snippetFallback;
+    }
+    if (fieldResultSetting.snippetSize) {
+      serverFieldResultSetting.snippet.size = fieldResultSetting.snippetSize;
+    }
+  }
+
+  return serverFieldResultSetting;
+};
+
+export const splitResultFields = (resultFields: FieldResultSettingObject, schema: Schema) => {
+  const textResultFields: FieldResultSettingObject = {};
+  const nonTextResultFields: FieldResultSettingObject = {};
+  const keys = Object.keys(schema);
+  keys.forEach((fieldName) => {
+    (schema[fieldName] === 'text' ? textResultFields : nonTextResultFields)[fieldName] =
+      resultFields[fieldName];
+  });
+
+  return { textResultFields, nonTextResultFields };
+};
+
+const convertToFieldResultSetting = (serverFieldResultSetting: ServerFieldResultSetting) => {
+  const fieldResultSetting: FieldResultSetting = {
+    raw: !!serverFieldResultSetting.raw,
+    snippet: !!serverFieldResultSetting.snippet,
+    snippetFallback: !!(
+      serverFieldResultSetting.snippet &&
+      typeof serverFieldResultSetting.snippet === 'object' &&
+      serverFieldResultSetting.snippet.fallback
+    ),
+  };
+
+  if (
+    serverFieldResultSetting.raw &&
+    typeof serverFieldResultSetting.raw === 'object' &&
+    serverFieldResultSetting.raw.size
+  ) {
+    fieldResultSetting.rawSize = serverFieldResultSetting.raw.size;
+  }
+
+  if (
+    serverFieldResultSetting.snippet &&
+    typeof serverFieldResultSetting.snippet === 'object' &&
+    serverFieldResultSetting.snippet.size
+  ) {
+    fieldResultSetting.snippetSize = serverFieldResultSetting.snippet.size;
+  }
+
+  return fieldResultSetting;
+};
+
+export const convertServerResultFieldsToResultFields = (
```

Comment:
[NOT a change request, just me thinking wistfully out loud]

If we have to have different client vs server data structures, I still think the best place to put that data transformation is in our `server/` route logic which can purely handle transforming JSON data between structure our APIs want vs structure that's easier to handle in the client-side.

It cleans up a _lot_ of mental overhead in the client side when there's only one data structure to expect/ deal with, and it just kinda makes sense to put it in server.

That being said, seriously don't do it now lol, but would love to hear what you think about the concept of it / me potentially moving these `convert` FNs to `server/` someday

Reply:
You're right, there would be some advantages to that. I could definitely be convinced.

However, I do see some advantages of converting client-side:

1. You can abstract the convert logic away client side in such a way that it still reduces logical overhead.
2. It's much easier to debug when the conversion happens client side, like you can see the network calls easily via the browser's Network tab and set breakpoints in the convert logic, etc. (Not that you can't do that server side, it just takes a bit more effort).
3. It also simplified the architecture to keep the Kibana server as a pass through, IMO. Which is simpler... a client that talks to a server that talks to a server? Or a client that talks to a single server? Like right now I think we can largely defer validation to ent-search, and we don't have to do a whole lot in the Kibana server. If we add conversion code to the server, proper validation does now matter.

That's just my 2 cents.